### PR TITLE
Introduce usdp gauge

### PR DESCRIPTION
**Pull request by 34x4p08**

### CODE REVIEW

File: contracts/USDPGauge.sol [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        if (block.timestamp >= periodFinish) {
+            rewardRate = reward.div(DURATION);
+        } else {
+            uint remaining = periodFinish.sub(block.timestamp);
+            uint leftover = remaining.mul(rewardRate);
+            rewardRate = reward.add(leftover).div(DURATION);
+        }
+        lastUpdateTime = block.timestamp;
+        periodFinish = block.timestamp.add(DURATION);
+        usdp.safeTransferFrom(owner(), address(this), reward);
```
Comment (Eenae) [inline contracts/USDPGauge.sol:176]: Are you sure it's `owner()`, not `msg.sender`?

**Review by Eenae — CHANGES_REQUESTED** [review]
Is there a test or migration wiring the contracts together? So that I can check the wiring.

**Comment by Eenae** [comment]
Also I suggest renaming `USDPGauge` to USDP staking, 
and making DUCK/USDP ratios configurable: `% of liquidationFee to qDuck`, `% of stabilityFee to qDuck`.

**Review by Eenae — CHANGES_REQUESTED** [review]

**Comment by Eenae** [comment]
Interest capitalization function could be implemented, which is immediate gas-efficient depositing of claimed rewards.

File: contracts/Foundation.sol [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+contract Foundation is IFoundation {
+  using SafeMath for uint;
+
+  IERC20 public constant usdp = IERC20(0x1456688345527bE1f37E9e627DA0837D6f08C925);
+
+  // TODO: set the address
+  address public constant qDuckDistribution = address(0);
+  // TODO: set the address
+  address public constant usdpGauge = address(0);
+  address public auction;
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/Foundation.sol]: Added gov control over this

File: contracts/Foundation.sol [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  function distribute() external override {
+    uint usdpBalance = usdp.balanceOf(address(this));
+
+    uint usdpGaugeAmount = (usdpBalance.add(liquidationFee)).div(2);
+    uint qDuckDistributionAmount = usdpBalance.sub(usdpGaugeAmount);
+
+    liquidationFee = 0;
+
+    usdp.transfer(usdpGauge, usdpGaugeAmount);
+    usdp.transfer(qDuckDistribution, qDuckDistributionAmount);
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/Foundation.sol]: acknowledged

File: contracts/Foundation.sol [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  function submitLiquidationFee(uint fee) external override auctionOnly {
+    liquidationFee = liquidationFee.add(fee);
+  }
+
+  function distribute() external override {
+    uint usdpBalance = usdp.balanceOf(address(this));
+
+    uint usdpGaugeAmount = (usdpBalance.add(liquidationFee)).div(2);
+    uint qDuckDistributionAmount = usdpBalance.sub(usdpGaugeAmount);
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/Foundation.sol]: Fixed the calculation

File: contracts/StabilityFeeCollector.sol [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "./interfaces/IUSDPGauge.sol";
+
+contract StabilityFeeCollector {
+
+  IERC20 public constant usdp = IERC20(0x1456688345527bE1f37E9e627DA0837D6f08C925);
+
+  // TODO: set the address
+  IUSDPGauge public constant usdpGauge = IUSDPGauge(address(0));
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/StabilityFeeCollector.sol]: Test case added

File: contracts/USDPGauge.sol [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    function withdraw(uint amount) public virtual {
+        _burn(msg.sender, amount);
+        usdp.safeTransfer(msg.sender, amount);
+    }
+}
+
+contract USDPGauge is TokenWrapper, Ownable {
+    using SafeMath for uint;
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/USDPGauge.sol]: applied

File: contracts/USDPGauge.sol (hunk 1 above)
Comment (34x4p08) [reply contracts/USDPGauge.sol]: fixed

File: contracts/USDPGauge.sol [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    using SafeERC20 for IERC20;
+    using Address for address;
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/USDPGauge.sol]: Cleaned up

File: contracts/USDPGauge.sol [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+    function multiClaim(address[] calldata users) public {
+        for (uint i; i < users.length; i++) {
+            _claim(users[i]);
+        }
+    }
+    
+    function _claim(address user) internal updateReward(user) {
+        require(user != address(0), "ZERO_ADDRESS");
+        uint reward = earned(user);
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/USDPGauge.sol]: applied

File: contracts/USDPGauge.sol [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        }
+    }
+
+    function lastTimeRewardApplicable() public view returns (uint) {
+        return Math.min(block.timestamp, periodFinish);
+    }
+
+    function rewardPerToken() public view returns (uint) {
+        if (totalSupply() == 0) {
+            return rewardPerTokenStored;
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/USDPGauge.sol]: Can be resolved by making a pre-deposit

File: contracts/USDPGauge.sol [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+        for (uint i; i < users.length; i++) {
+            _claim(users[i]);
+        }
+    }
+    
+    function _claim(address user) internal updateReward(user) {
+        require(user != address(0), "ZERO_ADDRESS");
+        uint reward = earned(user);
+        if (reward > 0) {
+            rewards[user] = 0;
```
[the remark this answers is not in the document]
Comment (34x4p08) [reply contracts/USDPGauge.sol]: applied

**Comment by 34x4p08** [comment]
> Also I suggest renaming `USDPGauge` to USDP staking,
> and making DUCK/USDP ratios configurable: `% of liquidationFee to qDuck`, `% of stabilityFee to qDuck`.
> Interest capitalization function could be implemented, which is immediate gas-efficient depositing of claimed rewards.

Implemented

**Review by Eenae — APPROVED** [review]

**Pull request merged**